mgr/volumes: Implement subvolume version v2

ShyamsundarR opened the pull request:
v2 subvolumes add the ability to retain snapshots post subvolume deletion.

Towards this, the following changes are made to support v2 subvolumes,
- Snapshots are created at the subvolume directory, rather than at the user data directory
- On subvolume rm operation, if --retain-snapshots option is used, current subvolume incarnation is deleted but snapshots are retained
- Added supported features for a subvolume in the info output, to make the ability to retain snapshots discoverable by consumers
- Added subvolume state information in the info output, to enable consumers to discover state of a subvolume in its retained or complete form
- Ability to auto upgrade to v2 subvolumes from v1 subvolumes if feasible, in the discover path

Other changes done to enable cleaner v2 implementation,
- Rejigged state machine implementation to make the state machine more comprehensible and easier to extend
- Added operation type based subvolume operations allowed during open, to decrease extending options to the subvolume open call, and also prevents abstraction leaking to other classes

Fixes: https://tracker.ceph.com/issues/45729
Signed-off-by: Shyamsundar Ranganathan srangana@redhat.com

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py:
```diff
@@ -42,7 +46,14 @@ def refresh(self):
         try:
             log.debug("opening config {0}".format(self.config_path))
-            fd = self.fs.open(self.config_path, os.O_RDONLY)
+            target_path = self.config_path
```
I am leaning towards not maintaining a per incarnation .meta.<UUID> file. The information can be retained in the .meta file as such, and snapshots would retain the .meta state as when it was taken, to reference the incarnation ID.

Addition of the symlink only causes some twisted code to maintain the same.

ShyamsundarR replied:
> Couple of points (and questions):
> 
>     * How would (and where) clone state (clone record in parent subvolume) be recorded if per-incarnation metadata is not maintained?

(also answering/addressing/pondering the bullets further below in the comment)

It is maintained in the .meta file, under the section "clone snaps" as before. The data in this section already encodes the (snapname, track-id) tuple, which gives the required path to access the .meta file within the snapshot to determine the incarnation the snapshot belongs to.

**NOTE:** The one place where we need to construct the snapshot incarnation directory to clone from, is handled in the function `subvolume_snapshot_path` which uses the .meta file within the snapshot to determine its subvolume incarnation path.

On a `rm` with `--retain-snaps` the .meta file global section is updated, as needed, to reflect that there is no current incarnation (I am being purposefully vague here about how the global section is updated to reflect the same, as we could use state/path/incarnation to do this, as is being discussed in other comments). All other sections in the .meta file are retained, including "protected snaps" and "clone snaps" sections.

When the last snapshot of a subvolume with no current incarnation is deleted, the subvolume is also deleted. (this is not coded yet)

My question/confusion is why do we need the "clone snaps" section recorded per-incarnation? I fail to understand that.

> I am not _really_ against keeping a single metadata instance as long as:
> 
>     * subvolume incarnations can be isolated to the versions they were created
> 
>     * can handle a mix of subvolume incarnations with and without `--retain-snapshots`
> 
>     * clone records can persist across subvolume incarnations
> 
> 
> Having symbolic links is a side-effect of using filesystem scratch space for metadata and to speed-up lookup operations for current incarnation.

vshankar replied:
> My question/confusion is why do we need the "clone snaps" section recorded per-incarnation? I fail to understand that.
> 
> >

Technically, right now, there is no hard requirement for 'clone snaps' section to be recorded in per-incarnation metadata -- or any other metadata for that matter -- once a subvolume is deleted, metadata that needs to be recorded for that incarnation can be stashed in one single .meta (shared section in metadata file). Will it work -- sure it will. Is it a clean design -- I don't think so -- because the (deleted) subvolume has certain parts of its metadata (read-only) in the snapshot and other metadata (read+write) outside the snapshot. Also, what if the (writable) metadata is subvolume version dependent? One can figure out the version from the metadata in the snapshot. However, when we move towards storing subvolume metadata in a key/value store, there won't be any metadata in the snapshot to figure out which version a (deleted) subvolume got created.

ShyamsundarR replied:
> (follow up question)
> 
> How does an incarnated subvolume relate its metadata when writable metadata (post subvolume deletion) is stashed in a place that's shared with other incarnations (including current, if any)?

Incarnations are properties of the same subvoume, the subvolume is the primary construct. Hence subvolume metadata having information regarding different incarnations is expected and not a violation.

> Right now, there is a 1:1 map between a subvolume version and its metadata, i.e., a given version is fully capable of understanding its metadata format. Sharing metadata pieces between incarnations seems to introduce unnecessary complications. E.g., say, a snapshot exist for subvolume incarnation 1 with version 2 and the current incarnation is 10 with version 5 -- stashing version 2 metadata in version 5 format does not look clean. Moreover, how does all of this tie-up with implementation? Should a version 2 subvolume _know_ how to persist metadata in (possibly _all_) other versions, or, should a version 5 subvolume _know_ how to deal with version 2 metadata? Right now, the metadata we are discussing is generic across versions. However, this _might_ not be the case going forward. In short, the question that we need to ask ourselves is: do we see a need to have metadata that is version dependent going forward?

Incarnations are not different versions. The version is tied to the subvolume, hence I do not see the above problem occurring.

If a version-n subvolume is retained, then its metadata and other on disk representations need to be upgraded to version-n++. Till such a time it remains as a version-n subvolume, and when upgraded it will not have any version-n subparts in its metadata.

Under these conditions, it is not feasible for a version-5 subvolume to have version-2 subparts in it, or a version-2 subvolume to have a version-5 incarnation created within it.

> 
> Also, how does all of this work with snapshot restore operation that will come up sooner or later? Does persisting both "clone source" and "clone record" in the same metadata with possibly different versions sound comprehensible? And what if the restore operation fails? Cleaning that up (removing half-baked subvolume) will not be a plain asynchronous purge.

Incarnations are not different versions, as laid out above.

> 
> IMO, this is leading to a considerable amount of complexity for a small win in dealing with a single instance (and possibly bloated) metadata chunk.

I think there is a gap here regarding how we perceive retained subvolumes. IMO, A retained subvolume that is recreated in the future is the same subvolume, hence its version/metadata are properties of the subvolume and not of its incarnations. Incarnations are way to separate retained-recreated chain of events on such a subvolme.

vshankar replied:
> Incarnations are not different versions. The version is tied to the subvolume, hence I do not see the above problem occurring.
> 
> If a version-n subvolume is retained, then its metadata and other on disk representations need to be upgraded to version-n++. Till such a time it remains as a version-n subvolume, and when upgraded it will not have any version-n subparts in its metadata.

mgr/volumes provisions subvolumes with the highest subvolume version it understands. A subvolume being retained can be version X and then recreated with version Y.

vshankar replied:
.. and BTW, who and when does the version "upgrade"?

ShyamsundarR replied:
> > Incarnations are not different versions. The version is tied to the subvolume, hence I do not see the above problem occurring.
> > If a version-n subvolume is retained, then its metadata and other on disk representations need to be upgraded to version-n++. Till such a time it remains as a version-n subvolume, and when upgraded it will not have any version-n subparts in its metadata.
> 
> mgr/volumes provisions subvolumes with the highest subvolume version it understands. A subvolume being retained can be version X and then recreated with version Y.

Good catch, this has to be fixed (in this PR) to use the current version subvolumes create/create_clone method (if a current version for the subvolume exists).

ShyamsundarR replied:
> .. and BTW, who and when does the version "upgrade"?

I would expect an auto-upgrade when possible (like legacy to v1, or v1 without snapshots to v2), as the first method.

Where auto upgrade to a particular future version is not possible, the user would need to upgrade the same (using any provided steps/method for the upgrade).

vshankar replied:
Why not allow (re)creating subvolumes with the current version? Is that something specific to CSI?

I think allowing subvolumes to be recreated with current (available) version allows for an easier "upgrade" path -- take a snapshot, restore the subvolume from the snapshot (which is a rm+clone)

vshankar replied:
> I would expect an auto-upgrade when possible (like legacy to v1, or v1 without snapshots to v2), as the first method.

The auto-upgrade path in mgr/volumes is for existing subvolumes to use features such as clone. We should generally avoid auto-uprgade on access kind of things (N-1 -> N or worse N-2 -> N). Moreover, IMO the recommended way to upgrade would be snap+restore. That requires maintaining multiple versions of a subvolume during upgrade, right?

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py:
```diff
@@ -70,7 +81,21 @@ def flush(self):
         try:
-            fd = self.fs.open(self.config_path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, self.mode)
+            if (incarnation != b""):
+                if (incarnation != MetadataManager.EMPTY_INCARNATION):
```
For a snapshot preserved volume, we would still leave behind a .meta with a volume state that is "empty"

ShyamsundarR replied:
Incarnation is not carried over to the metadata file any longer.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:
```diff
@@ -99,6 +108,73 @@ def load_config(self):
+                e = VolumeException(-e.args[0], e.args[1])
+            raise e
+
+    def add_clone_source(self, volname, subvolume, snapname, flush=False):
```
As most of the implementation is common across v1/v2 subvolume versions, the code would move to the subvolume base file, and where appropriate retain the base class method with the same name.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/volume.py:
```diff
@@ -193,7 +197,7 @@ def resize_subvolume(self, **kwargs):
             with open_volume(self, volname) as fs_handle:
                 with open_group(fs_handle, self.volspec, groupname) as group:
-                    with open_subvol(fs_handle, self.volspec, group, subvolname) as subvolume:
+                    with open_subvol(fs_handle, self.volspec, group, subvolname, need_incarnation=True) as subvolume:
```
Operations needing an incarnation are flagged of using the "need_incarnation" flag as in here.

ShyamsundarR replied:
Current implementation uses the op_type and lets the subvolume version decide allowed states and types.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/op_sm.py:
```diff
@@ -10,15 +10,19 @@ class OpSm(object):
     FAILED_STATE = 'failed'
     FINAL_STATE  = 'complete'
     CANCEL_STATE = 'canceled'
+    SNAP_RETAINED_STATE = 'snap-retained'
```
(providing an over-arching discussion for comments around state, incarnation and their values)

A subvolume with no incarnation (but existing snapshots), needs a manner to determine that it is in the `no-incarnation` state. This is needed as, some operations are allowed on these subvolumes and some others not.

Here are the operation classifications for a subvolume in the `no-incarnation` state,
- Allowed operations: create, snapshot ls, snapshot rm, snapshot protect, snapshot unprotect, snapshot clone
- Disallowed operations: rm, getpath, snapshot create, resize
- Operations being mused about: ls, info

The options to determine `no-incarnation` state are,
1) Leave the .meta file as is and determine incarnation is missing/deleted based on `path`, from the global section, not being found
- This is implicit, and needs a stat of the path and handling the same
- non-optimal and hence dropped

2) Change `path` in global section, to an empty string or remove `path` option from the global section
- more explicit, and does not need any further inspection of the subvolume

  I felt uncomfortable leaving the subvolume in the `complete` state in the .meta file, hence instead of the above,
3) Change `state` to denote a subvolume in the no incarnation state
- `path` value really does not matter, but again leaving it as is, can be misleading, so (2) needs to be done anyway

As long as we understand 'complete' state, as the subvolume is ready for certain operations, which includes it possibly being in `no-incarnation` state as well, we need not do (3) and can just stick to (2).

The addition of `incarnation` in the global section instead of dissecting the path to determine the same, was done as I felt this needed to be explicit than derived from path. This hence led to `EMPTY_INCARNATION` as well (or, removing the `incarnation` key from the .meta file). For v2 subvolumes, as long as the understanding is that the last component of the `path` *IS* the incarnation, we need not store the same explicitly.

If we agree to option (2) above for `no-incarnation` sub-state and that we can extract the `incarnation` from the `path` where it exists, the comments surrounding these changes can be addressed appropriately.

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -0,0 +1,311 @@
+            log.warning("error updating protected snap list ({0})".format(me))
+            raise VolumeException(-errno.EINVAL, "error unprotecting snapshot")
+
+    def protect_snapshot(self, snapname):
```
I do not see a need to have protect/unprotect calls for v2 subvolumes.

ShyamsundarR replied:
Removed as part of https://github.com/ceph/ceph/pull/35755

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py:
```diff
@@ -42,7 +46,14 @@ def refresh(self):
         conf_data = StringIO()
         try:
             log.debug("opening config {0}".format(self.config_path))
```
Dropping a parallel note, on snapshot incarnation retention, based on the conversation in this [comment chain](https://github.com/ceph/ceph/pull/35647/files#r442168572), as shared with @vshankar and @kotreshhr 

**Problem:** When moving subvolume metadata information stored as files within the subvolume, and hence automatically within each of its snapshots, to a key-value store, we cannot rely on the subvolume snapshot carrying a metadata copy as a reference to its incarnation.

This leads to the following requirements.

**Requirements:**
a. Given a (subvolname, snapname) tuple, its incarnation information should be available
- NOTE: It is assumed, that (volume, subvolumegroup) is part of the tuple, but not mentioned explicitly
- For v1 subvolumes, this comes from the subvolume path key
- For v2 subvolumes, this information needs to be stored explicitly, as the current subvolume incarnation/path need not be the same as when the snapshot was created

b. All state data updates, regarding the subvolume-snapshots, should be operated on from a single source
- This information, at present, would include,
  - Snapshots protected
  - Snapshots being cloned

**Current solution for (a):**
- Subvolume snapshot metadata file as a wormhole to its incarnation
  - Currently implemented by reading the incarnation information from the metadata file directly within the subvolume snapshot
  - Alternately, using symlink information of the metadata file within the subvolume snapshot, to read from an incarnation specific metadata file at the subvolume
- Both approaches rely on the snapshot storing a metadata file for reference to its incarnation, and hence, when moving to a key-value store, makes the scheme non-future proof

**Alternate solution(s) for (a):**
- Implicitly infer the snapshot incarnation by inspecting the snapshot directory listing to determine the path
- Explicitly store per snapshot its incarnation information in the meta file and later in the key-value store, in a separate "snapshots" section

**For requirement (b):**
- All state data is already stored and operated on, from the primary metadata file of the subvolume
  - When a subvolume is deleted, but its snapshots retained, this primary metadata files "global' section is only updated as needed, and other sections are retained as is
  - The metadata file within the snapshot is referred to only for the incarnation information (which is being dealt with as part of (a) above)
- This hence could move seamlessly to the key-value store in the future

ShyamsundarR replied:
> **Alternate solution(s) for (a):**
> 
>     * Implicitly infer the snapshot incarnation by inspecting the snapshot directory listing to determine the path

This alternative is implemented.

> **For requirement (b):**
> 
>     * All state data is already stored and operated on, from the primary metadata file of the subvolume
>       
>       * When a subvolume is deleted, but its snapshots retained, this primary metadata files "global' section is only updated as needed, and other sections are retained as is
>       * The metadata file within the snapshot is referred to only for the incarnation information (which is being dealt with as part of (a) above)
> 
>     * This hence could move seamlessly to the key-value store in the future

Single meta file is retained with prior metadata.

ShyamsundarR commented:
@vshankar @kotreshhr @ajarr redid the implementation to handle the following as per discussions on single .meta file and subvolume being retained without mixed incarnation versions.

- Updated the state machine, with the concrete example provided by @vshankar Thanks!
- Made changes to use operation type, hence reduced per operation flags to open

The above two are in the first commit (apologies, one of the last rebases locally merged the two in my branch, else were retained separately)

- Updated the code for handling v2 subvolumes

I will be following this up with a commit that moves common code between v1/2 to the base class, and then the rest of the TODOs to keep commits separate and possibly easier to review.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/async_cloner.py:
```diff
@@ -10,12 +10,17 @@
 from .operations.subvolume import open_subvol
 from .operations.clone_index import open_clone_index
+from .operations.template import SubvolumeOpType
+from .operations.versions import SubvolumeBase
```
This import is not required.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/async_cloner.py:
```diff
@@ -10,12 +10,17 @@
+from .operations.op_sm import SubvolumeOpSm
+from .operations.op_sm import SubvolumeTypes
+from .operations.op_sm import SubvolumeActions
+from .operations.op_sm import SubvolumeStates
```
Combine multiple imports form the same file into a single line

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -160,39 +196,53 @@ def _get_clone_source(self):
         subvol_path = self.path
         return self._resize(subvol_path, newsize, noshrink)
 
+    def info(self):
```
Should the pattern be to define `info` in the base class, and not implement the method in v1, and use it in v2 appropriately calling it via `super(SubvolumeBase, self).info`?

ShyamsundarR replied:
Updated with the super pattern.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/__init__.py:
```diff
@@ -44,24 +48,43 @@ def _get_subvolume_version(self, version):
+        if version >= SubvolumeV2.version():
+            return
+
+        v1_subvolume = self._get_subvolume_version(version)(subvolume.fs, subvolume.vol_spec, subvolume.group, subvolume.subvolname)
```
Should attempt opening the v1 subvolume and handle not complete errors, before proceeding (in flight clones, or clone failed/cancelled state subvolumes should not be upgraded as yet).

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -0,0 +1,311 @@
+    def open(self, need_complete=True, expected_types=[], need_incarnation=False):
+        try:
+            self.metadata_mgr.refresh()
+            incarnation = self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_INCARNATION).encode('utf-8')
```
Incarnation is not carried over to the metadata file any longer.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/op_sm.py:
```diff
@@ -53,16 +61,19 @@ def get_init_state(stm_type):
 
     @staticmethod
-    def get_next_state(stm_type, current_state, ret):
+    def get_next_state(stm_type, current_state, event):
```
OpSm redone based on the concrete example provided.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/subvolume.py:
```diff
@@ -57,13 +57,17 @@ def remove_subvol(fs, vol_spec, group, subvolname, force=False):
     nc_flag = True if not force else False
-    with open_subvol(fs, vol_spec, group, subvolname, need_complete=nc_flag) as subvolume:
+    snapspresent = False
+    with open_subvol(fs, vol_spec, group, subvolname, need_complete=nc_flag, need_incarnation=True) as subvolume:
```
Current implementation uses the op_type and lets the subvolume version decide allowed states and types.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py:
```diff
@@ -25,8 +26,11 @@ class MetadataManager(object):
 
     MAX_IO_BYTES = 8 * 1024
+    PATH_MAX = 4096
+    EMPTY_INCARNATION = b"empty"
```
Incarnation is no longer used as part of the metadata.

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -0,0 +1,418 @@
+            # persist subvolume metadata
+            qpath = subvol_path.decode('utf-8')
+            try:
+                self.metadata_mgr.refresh()
```
Would it help if we reach here after an open() design wise? So the metadata is already readout.

ShyamsundarR replied:
The issue that I see is, we would need to expose state to [`create_subvolume`](https://github.com/ceph/ceph/blob/7fb21a25e1cdcce27af0f0f9b040aa1537fb42fc/src/pybind/mgr/volumes/fs/volume.py#L160-L168), such that it can decide to call create or work with the opened subvolume on open success. Similarly for `clone_subvolume_snapshot`.

It looks better to abstract away this detail for users of Subvolume*.

If we decide to add "--force", then `create_subvolume` can decide based on open and state if this should be allowed, but again this can also be abstracted away to the Subvolume implementation, to fail the Subvolume.create if "--force" is not used.

Thoughts?

vshankar replied:
> The issue that I see is, we would need to expose state to [`create_subvolume`](https://github.com/ceph/ceph/blob/7fb21a25e1cdcce27af0f0f9b040aa1537fb42fc/src/pybind/mgr/volumes/fs/volume.py#L160-L168), such that it can decide to call create or work with the opened subvolume on open success. Similarly for `clone_subvolume_snapshot`.

state need not be passed back to create_subvol() -- just that create() would be valid on a valid subvolume object.

Would that help with the resize case where attr changes can be done? That could possibly reduce abstract methods such as resize, etc..

ShyamsundarR replied:
open() returning a success for both ENOENT and EEXIST would help, with create picking up the idempotent requirements. IOW, create() handles updating attrs on an existing subvolume or creating one afresh.

Although I would not want to add it to this PR, considering its current size and scope.

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -0,0 +1,418 @@
+        except ValueError:
+            return False
+
+    def snapshot_data_path(self, snapname):
```
Design-wise accessing snapshots does not require a valid subvolume object. Should snapshots be treated as a separate entity altogether?

ShyamsundarR replied:
Snapshots in V1 depend on the path component from the meta file, in V2 are not dependent on the meta file for lookups.

Snapshots are dependent on updating the meta file for attach/detach operations as such.

As the meta file is managed via the subvolume, and V2 is derived from V1, I would leave snapshots as is for now.

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/__init__.py:
```diff
@@ -81,7 +81,7 @@ def upgrade_legacy_subvolume(self, fs, subvolume):
 
     def get_subvolume_object(self, fs, vol_spec, group, subvolname, upgrade=True):
-        subvolume = SubvolumeBase(fs, vol_spec, group, subvolname)
+        subvolume = SubvolumeDiscover(fs, vol_spec, group, subvolname)
```
Why carve this out into a separate class?

(also, SubvolumeDiscover does not have abstract discover method)

ShyamsundarR replied:
> Why carve this out into a separate class?

The intention was to provide abstract methods in the base class, hence it renders creating an instance of the base class invalid, thus leading to a derived class for help with discovery.

> (also, SubvolumeDiscover does not have abstract discover method)

Do you mean we should move discover into this class and make it abstract in the base class? Which sounds like a good idea to me. (currently discover is not an abstract method)

vshankar replied:
> Do you mean we should move discover into this class and make it abstract in the base class? Which sounds like a good idea to me. (currently discover is not an abstract method)

Right, I expected discover() as an abstract method. Thinking more -- can upgrade() be an abstact method that can be invoked to upgrade a subvolume to the latest version?

ShyamsundarR replied:
Now that V2 is derived from V1, the need for an abstract base class is not present any longer. Hence SubvolumeDiscover is removed.

I did play around with an upgrade method briefly earlier, but it did not help much. Let's leave the upgrade explicit during the `get_subvolume_object` process.

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:
```diff
@@ -304,3 +334,130 @@ def _info (self):
             'bytes_quota': "infinite" if nsize == 0 else nsize, 'bytes_used': int(usedbytes),
             'bytes_pcent': "undefined" if nsize == 0 else '{0:.2f}'.format((float(usedbytes) / nsize) * 100.0),
             'pool_namespace': pool_namespace, 'features': features, 'state': state}
+
```
I think this change would not be required is V2 derives from V1. Implementations are not drastically different for the most part.

ShyamsundarR replied:
Applies to [this](https://github.com/ceph/ceph/pull/35647#discussion_r455046581) comment as well.

I decided to move the methods to the baseclass based on the public methods of the base class itself, and let the derived classes override any specifics. Further declaring abstract methods in the base class to help things along for v3, or at least making it more explicit on what derived classes should implement.
NOTE: I should probably add abstract methods for create/open/create_clone as these are missing in the base class, if we continue along this line.

This also helped me rationalize the code more easily, as either defined in base class or in the derived class.

I did not think about deriving v1 from v2, and agree that the move of methods to base class from v1 can be avoided, if we derive v2 from v1.

If it is mandatory or correct that v2 be derived from v1 I can work on that. If it is a matter of choice and either scheme is acceptable, I would like to retain the current one as I find it personally easier to rationalize.

vshankar replied:
> If it is mandatory or correct that v2 be derived from v1 I can work on that. If it is a matter of choice and either scheme is acceptable, I would like to retain the current one as I find it personally easier to rationalize.

It is certainly not mandatory, though, I recommend that we go this route.

ShyamsundarR replied:
In lieu of overall code churn due to this patch, moved to deriving V2 from V1, to reduce code movement from V1 into base class.

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -30,18 +22,19 @@ class SubvolumeV1(SubvolumeBase, SubvolumeTemplate):
+    def snapshot_path(self, snapname):
+        return os.path.join(self.snapshot_base_path(), snapname.encode('utf-8'))
+
+    def snapshot_data_path(self, snapname):
```
can you explain a bit more about snapshot_data_path() in comparison to snapshot_path()?

ShyamsundarR replied:
NOTE (digression): I should add (in the doc commit) comments on what is a v1/2 subvolume and the explanation that suits various snapshot paths as per this discussion

Initial comments on the various snapshot paths are provided [here](https://github.com/ceph/ceph/pull/35647/files#diff-a46bf7683125b6992877591c70815e01R115-R128) and I am going to build on these.

snapshot_base_path(self): Base path for all snapshots, for a subvolume
snapshot_path(snapname): Path to a specific snapshot named 'snapname', for a subvolume
snapshot_data_path(snapname): Path to user data directory within a subvolume snapshot named 'snapname'

snapshot_path is a path to a CephFS snapshot named snapname under the subvolume, and snapshot_data_path is the path within a CephFS snapshot under a subvolume, that contains user data.

As you know, in v1 both paths are the same. In v2, they are different as the data_path is the incarnation directory within the CephFS snapshot.

:+1: If a better name suits snapshot_data_path, will gladly change it.

vshankar replied:
Thanks for explaining -- I'll relook at the snapshot_data_path() usage.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/async_cloner.py:
```diff
@@ -10,12 +10,17 @@
 from .exception import IndexException, MetadataMgrException, OpSmException, VolumeException
 from .fs_util import copy_file
-from .operations.op_sm import OpSm
+from .operations.op_sm import SubvolumeOpSm
```
Fixed as part of later commits.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/op_sm.py:
```diff
@@ -1,68 +1,132 @@
+        raise OpSmException(-errno.EINVAL, "invalid state '{0}'".format(value))
+
+@unique
+class SubvolumeActions(Enum):
```
These are used outside the subvolume set of classes by the async_cloner. Hence it would be better to retain the prefix so as to be explicit in async_cloner.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:
```diff
@@ -242,24 +255,29 @@ def discover(self):
-    def info (self):
-        subvolpath = self.metadata_mgr.get_global_option('path')
-        etype = SubvolumeTypes.from_value(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_TYPE))
+    def _info (self):
```
Thanks. Will amend to the pattern.

ShyamsundarR replied:
Done

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -0,0 +1,418 @@
+        except (VolumeException, MetadataMgrException, cephfs.Error) as e:
+            try:
+                log.info("cleaning up subvolume with path: {0}".format(self.subvolname))
+                self.remove()
```
:+1: Yes.
This was one of the TODOs that I just pushed out yesterday (see commit: https://github.com/ceph/ceph/pull/35647/commits/150d05815293193ac7114a0a83eace2457380bed )

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:
```diff
@@ -277,9 +295,12 @@ def info (self):
             'bytes_quota': "infinite" if nsize == 0 else nsize, 'bytes_used': int(usedbytes),
             'bytes_pcent': "undefined" if nsize == 0 else '{0:.2f}'.format((float(usedbytes) / nsize) * 100.0),
-            'pool_namespace': pool_namespace, 'features': self.features}
+            'pool_namespace': pool_namespace, 'features': features, 'state': state}
```
Addressed as a part of later commits.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/async_cloner.py:
```diff
@@ -67,9 +70,13 @@ def get_clone_source(clone_subvolume):
     try:
         if should_cancel():
-            next_state = OpSm.get_next_state("clone", "pending", -errno.EINTR)
+            next_state = SubvolumeOpSm.transition(SubvolumeTypes.TYPE_CLONE,
```
I can see an errno based helper replace code at line 170 below (try/catch block).

In places where we decide based on cancel status, the errno is passed explicitly. Calling the helper based on errno would be equivalent.

At present as the errno based state change pattern is in only one place (line 170), hence a helper is not (yet) required?

ShyamsundarR replied:
Done

ShyamsundarR commented:
Discussed the following potential TODOs in this PR with @vshankar  and @kotreshhr 

Decisions from the discussion is presented below,

1. Add --force to subvolume create to ensure recreation of a snapshot retained subvolume is an explicit user choice

Not required currently, it is an users choice to retain anyway and also consumers like CSI use UUIDs to create new subvolumes such that collision probabilities are low to non-existent.

2. Decide if subvolume rather than the incarnation should hold the size, nspace, pool attributes and incarnation should hold mode, uid, gid

The incarnation needs the quota and pool attributes, as we leverge (or plan to) the same from a snapshot when cloning the subvolume. Hence we will retain the attributes with the incarnation as it exists now.

3. Add snapshot path to snapshot info command

No real use case to provide this information as yet. Not being implemented as part of this PR hence

ajarr commented on src/pybind/mgr/volumes/fs/async_cloner.py:
```diff
@@ -52,6 +52,26 @@ def open_at_group(volume_client, fs_handle, groupname, subvolname, op_type):
+
+
+@contextmanager
+def open_clone_volume_pair(volume_client, fs_handle, volname, groupname, subvolname):
```
s/volume/subvolume/

ShyamsundarR replied:
Done

ajarr commented on src/pybind/mgr/volumes/fs/async_cloner.py:
```diff
@@ -147,44 +171,47 @@ def cptree(src_root_path, dst_root_path):
-                with open_at_group(volume_client, fs_handle, s_groupname, s_subvolname) as source_subvolume:
-                    source_subvolume.detach_snapshot(s_snapname, index)
-                    clone_subvolume.remove_clone_source(flush=True)
+            with open_clone_volume_pair(volume_client, fs_handle, volname, groupname, subvolname) as clone_volumes:
```
s/open_clone_volume_pair/open_clone_subvolume_pair/

ShyamsundarR replied:
Done

ajarr commented on src/pybind/mgr/volumes/fs/async_cloner.py:
```diff
@@ -147,44 +171,47 @@ def cptree(src_root_path, dst_root_path):
-                    source_subvolume.detach_snapshot(s_snapname, index)
-                    clone_subvolume.remove_clone_source(flush=True)
+            with open_clone_volume_pair(volume_client, fs_handle, volname, groupname, subvolname) as clone_volumes:
+                clone_volumes[1].detach_snapshot(clone_volumes[2], index)
```
s/clone_volumes/clone_subvolumes/

ShyamsundarR replied:
Done

ajarr commented on src/pybind/mgr/volumes/fs/async_cloner.py:
```diff
@@ -147,44 +171,47 @@ def cptree(src_root_path, dst_root_path):
-                    clone_subvolume.remove_clone_source(flush=True)
+            with open_clone_volume_pair(volume_client, fs_handle, volname, groupname, subvolname) as clone_volumes:
+                clone_volumes[1].detach_snapshot(clone_volumes[2], index)
+                clone_volumes[0].remove_clone_source(flush=True)
```
s/clone_volumes/clone_subvolumes/

ShyamsundarR replied:
Done

ajarr commented on src/pybind/mgr/volumes/fs/async_cloner.py:
```diff
@@ -52,6 +52,26 @@ def open_at_group(volume_client, fs_handle, groupname, subvolname, op_type):
+def open_clone_volume_pair(volume_client, fs_handle, volname, groupname, subvolname):
+    with open_at_group(volume_client, fs_handle, groupname, subvolname, SubvolumeOpType.CLONE_INTERNAL) as clone_subvolume:
+        s_volname, s_groupname, s_subvolname, s_snapname = get_clone_source(clone_subvolume)
+        if groupname == s_groupname and subvolname == s_subvolname:
```
when is destination subvolume as the source?

ajarr commented on src/pybind/mgr/volumes/fs/async_cloner.py:
```diff
@@ -52,6 +52,26 @@ def open_at_group(volume_client, fs_handle, groupname, subvolname, op_type):
 
+@contextmanager
+def open_at_group_unique(volume_client, fs_handle, s_groupname, s_subvolname, c_subvolume, c_groupname, c_subvolname, op_type):
+    if s_groupname == c_groupname and s_subvolname == c_subvolname:
```
when is destination subvolume as the source?

ShyamsundarR replied:
Added a comment explaining the same,
```
# if a snapshot of a retained subvolume is being cloned to recreate the same subvolume, return
# the clone subvolume as the source subvolume
```

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/op_sm.py:
```diff
@@ -10,15 +10,19 @@ class OpSm(object):
     OP_SM_SUBVOLUME = {
-        INIT_STATE_KEY : FINAL_STATE,
+        INIT_STATE_KEY  : FINAL_STATE,
+        FINAL_STATE     : ('', ('', '', SNAP_RETAINED_STATE))
```
Removed as part of https://github.com/ceph/ceph/pull/35755

ShyamsundarR replied:
Resolved in the later commits based on the concrete example you provided. Thanks!

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/op_sm.py:
```diff
@@ -0,0 +1,114 @@
+from ...exception import OpSmException
+from .subvolume_attrs import SubvolumeTypes, SubvolumeStates, SubvolumeActions
+
+class SubvolumeTransitionKey(object):
```
Done.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -115,21 +118,48 @@ def create_clone(self, pool, source_volname, source_subvolume, snapname):
+                SubvolumeOpType.CLONE_INTERNAL}
+
+    def open(self, op_type):
+        if not isinstance(op_type, SubvolumeOpType):
```
Done

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:
```diff
@@ -111,9 +112,21 @@ def legacy_mode(self):
+        raise VolumeException(-errno.EINVAL, "invalid to invoke base class method")
+
+    @property
+    def type(self):
```
Done

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/op_sm.py:
```diff
@@ -1,68 +1,132 @@
+    ACTION_FAILED       = 2
+    ACTION_CANCELLED    = 3
+
+class SubvolumeTransitionKey(object):
```
Done

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:
```diff
@@ -242,24 +255,29 @@ def discover(self):
-            log.info("subvolume with path '{0}' moved to trashcan".format(subvol_path))
+        self._trash_dir(self.base_path)
+
+    def trash_incarnation_dir(self):
```
Done

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -0,0 +1,418 @@
+
+log = logging.getLogger(__name__)
+
+class SubvolumeV2(SubvolumeBase, SubvolumeTemplate):
```
Done

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/subvolume.py:
```diff
@@ -26,6 +26,9 @@ def create_subvol(fs, vol_spec, group, subvolname, size, isolate_nspace, pool, m
     :param gid: the group identifier
     :return: None
     """
+    # NOTE: create/clone_create uses max version, as we get here either with a v2 (current max) retained subvolume, or a
```
Done

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -0,0 +1,418 @@
+        except cephfs.Error as e:
+            if e.errno == errno.ENOENT:
+                raise VolumeException(-errno.ENOENT,
+                                      "snapshot '{0}' doesnot exist".format(snapname))
```
Done

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -115,21 +118,48 @@ def create_clone(self, pool, source_volname, source_subvolume, snapname):
             raise e
 
-    def open(self, need_complete=True, expected_types=[]):
+    def allowed_ops_by_type(self, vol_type):
```
This is a useful construct to have, but again based on the size and scope of this PR, would like to pick this up subsequently and not as a part of this change.

kotreshhr commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -115,21 +118,48 @@ def create_clone(self, pool, source_volname, source_subvolume, snapname):
+
+            estate = self.state
+            if op_type not in self.allowed_ops_by_state(estate):
+                raise VolumeException(-errno.EAGAIN, "subvolume '{0}' is not ready for operation {1}".format(
```
Is not EAGAIN more apt for all non supported ops? In the usecase like subvolume with snapshot is deleted and a getpath on that subvolume should return ENOENT to user right ? Sorry if this is discussed before.

ShyamsundarR replied:
Done.

kotreshhr commented on src/pybind/mgr/volumes/module.py:
```diff
@@ -83,7 +83,8 @@ class Module(orchestrator.OrchestratorClientMixin, MgrModule):
-                   'name=force,type=CephBool,req=false ',
+                   'name=force,type=CephBool,req=false '
+                   'name=retain_snapshots,type=CephBool,req=false ',
             'desc': "Delete a CephFS subvolume in a volume, and optionally, "
```
nit: It's better to add retain snapshot feature in cli help ?

ShyamsundarR replied:
Done.

ShyamsundarR commented:
> Two generic questions with this PR.
> 
>     1. After subvolume deletion, subvolume ls still lists subvolume. Is this expected ?

Yes, initial thoughts were to not list the subvolume. During implementation found that listing and reflecting the status in info is better, as part of the subvolume is retained.

Once I add the doc commit to this, I will add clarification on info output as it pertains to retained subvolumes and the behavior in listing.

> 
>     2. The snapshot path is changed. How does it affect the user using older version and this version ? I remember we discussed this but didn't get how the snapshot path currently being used.

User with an older subvolume version and having snapshots will not be able to [`--retain-snapshots`](https://github.com/ceph/ceph/pull/35647/files?file-filters%5B%5D=.py&hide-deleted-files=true#diff-e150aca08e47cf42cf6103319b3a2f30R230-R231) on subvolume removal. Also, such V1 subvolumes, with snapshots, will not be [auto upgraded](https://github.com/ceph/ceph/pull/35647/files?file-filters%5B%5D=.py&hide-deleted-files=true#diff-a5ac45a80040e84707c64d528a685234R70-R72) for the user.

(related to listing as well) The subvolume info output details what is supported [based on](https://github.com/ceph/ceph/pull/35647/files?file-filters%5B%5D=.py&hide-deleted-files=true#diff-228e04ea5fdfd99eef3c5001f6d39c88R265) the state. So from a listing and info perspective, it is better to list and pass back information on retained subvolumes, as it aids discovery of features and state.

ajarr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1650,6 +1907,234 @@ def test_subvolume_rm_with_snapshots(self):
+        self._fs_cmd("subvolume", "snapshot", "rm", self.volname, subvolume, snapshot1)
+        self._fs_cmd("subvolume", "snapshot", "rm", self.volname, subvolume, snapshot2)
+        self._fs_cmd("subvolume", "snapshot", "rm", self.volname, clone2, snapshot3)
+
```
You can maybe `subvolume ls` and check whether they are removed?

ShyamsundarR replied:
Done, nice idea.

ShyamsundarR commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1622,6 +1677,208 @@ def test_subvolume_upgrade(self):
+
+        # ls
+        subvolumes = json.loads(self._fs_cmd('subvolume', 'ls', self.volname))
+        self._assert_meta_location_and_version(self.volname, subvolume, version=1)
```
Ack, testing once at the end, as you state, would suffice.

ShyamsundarR replied:
Done.

ShyamsundarR commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1650,6 +1907,234 @@ def test_subvolume_rm_with_snapshots(self):
+
+    def test_subvolume_retain_snapshot_clone(self):
+        """
+        series of test cases to clone from retained subvolume snapshots, ensure a clone chain works, ensure
```
:smiley: Ack! thought so as well. Will break this into smaller tests.

ShyamsundarR replied:
Done.

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -0,0 +1,418 @@
+
+        subvol_path = os.path.join(self.base_path, str(uuid.uuid4()).encode('utf-8'))
+        try:
+            # TODO: decide if subvolume rather than the incarnation should hold the size, nspace, pool attributes
```
Added comments for [V1](https://github.com/ceph/ceph/pull/35647/files#diff-e150aca08e47cf42cf6103319b3a2f30R25-R38) and [V2](https://github.com/ceph/ceph/pull/35647/files#diff-228e04ea5fdfd99eef3c5001f6d39c88R20-R38) subvolume, including incarnation and what it means in this context.

ShyamsundarR commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1622,6 +1677,208 @@ def test_subvolume_upgrade(self):
+        self.assertNotEqual(len(subvol_info), 0, "expected the 'fs subvolume info' command to list metadata of subvolume")
+        for md in subvol_md:
+            self.assertIn(md, subvol_info.keys(), "'{0}' key not present in metadata of subvolume".format(md))
+        self.assertNotEqual(subvol_info["bytes_quota"], "infinite", "bytes_quota should not be set to infinite if quota is not set")
```
Done (all instances)

ShyamsundarR commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1622,6 +1677,208 @@ def test_subvolume_upgrade(self):
+        # this would attempt auto-upgrade on access, but fail to do so
+        subvolpath = self._fs_cmd("subvolume", "getpath", self.volname, subvolume)
+        self._assert_meta_location_and_version(self.volname, subvolume, version=1)
+        self.assertNotEqual(subvolpath, None)
```
Done (all instances)

ShyamsundarR commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1622,6 +1677,208 @@ def test_subvolume_upgrade(self):
+        self._assert_meta_location_and_version(self.volname, subvolume, version=1)
+        self.assertNotEqual(len(snap_info), 0)
+        for md in snap_md:
+            self.assertIn(md, snap_info.keys(), "'{0}' key not present in metadata of snapshot".format(md))
```
Done (all instances)

ShyamsundarR commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1622,6 +1677,208 @@ def test_subvolume_upgrade(self):
+        # snap-info
+        snap_info = json.loads(self._get_subvolume_snapshot_info(self.volname, subvolume, snapshot))
+        self._assert_meta_location_and_version(self.volname, subvolume, version=1)
+        self.assertNotEqual(len(snap_info), 0)
```
Removed in all instances

ShyamsundarR commented:
@ajarr @vshankar @kotreshhr Added documentation.

ShyamsundarR commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -56,8 +59,15 @@ def _wait_for_clone_to_fail(self, clone, clone_group=None, timo=120):
 
-    def _verify_clone_attrs(self, subvolume, clone, source_group=None, clone_group=None):
-        path1 = self._get_subvolume_path(self.volname, subvolume, group_name=source_group)
+    def _get_subvolume_snapshot_path(self, snapshot, subvol_path):
```
When [this](https://github.com/ceph/ceph/pull/35756) PR is completed, version based handling of path to the snapshot would get taken care of.

kotreshhr commented:
> > Two generic questions with this PR.
> > ```
> > 1. After subvolume deletion, subvolume ls still lists subvolume. Is this expected ?
> > ```
> 
> Yes, initial thoughts were to not list the subvolume. During implementation found that listing and reflecting the status in info is better, as part of the subvolume is retained.
> 

   I think it's a good idea to mention 'only snapshots are retained' along with subvolume names for this case. Users can filter this out if not interested. With current output, it can't differentiated. What do you think?

kotreshhr commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -49,20 +42,42 @@ class SubvolumeV2(SubvolumeBase, SubvolumeTemplate):
+    def snapshot_data_path(self, snapname):
+        snap_base_path = self.snapshot_path(snapname)
+        uuid_str = None
+        with self.fs.opendir(snap_base_path) as dir_handle:
```
snapshot info on non-existing snapshot crashes here.

```
# bin/ceph fs subvolume snapshot info a sub_0 snap_1
Error EINVAL: Traceback (most recent call last):
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/mgr_module.py", line 1170, in _handle_command
    return self.handle_command(inbuf, cmd)
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/volumes/module.py", line 357, in handle_command
    return handler(inbuf, cmd)
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/volumes/module.py", line 471, in _cmd_fs_subvolume_snapshot_info
    group_name=cmd.get('group_name', None))
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/volumes/fs/volume.py", line 345, in subvolume_snapshot_info
    snap_info_dict = subvolume.snapshot_info(snapname)
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py", line 273, in snapshot_info
    snappath = self.snapshot_data_path(snapname)
  File "/root/sandbox/shyam-ceph/ceph/src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py", line 65, in snapshot_data_path
    with self.fs.opendir(snap_base_path) as dir_handle:
  File "cephfs.pyx", line 1073, in cephfs.LibCephFS.opendir
cephfs.ObjectNotFound: opendir failed: No such file or directory [Errno 2]

```

ShyamsundarR replied:
Good catch, will add a test case and also add exception handling within snapshot_data_path to only throw VolumeException on errors.

ShyamsundarR replied:
Fixed.

kotreshhr commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -0,0 +1,434 @@
+    def remove(self, retainsnaps=False):
+        if self.list_snapshots():
+            if not retainsnaps:
+                raise VolumeException(-errno.ENOTEMPTY, "subvolume '{0}' has snapshots".format(self.subvolname))
```
On V2, shouldn't this be ENOENT ?

```
# bin/ceph fs subvolume rm a sub_0
Error ENOTEMPTY: subvolume 'sub_0' has snapshots

```

kotreshhr commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -0,0 +1,434 @@
+    def remove(self, retainsnaps=False):
+        if self.list_snapshots():
+            if not retainsnaps:
+                raise VolumeException(-errno.ENOTEMPTY, "subvolume '{0}' has snapshots".format(self.subvolname))
```
This is correct errno when a removal of subvolume is issued which has snapshots and also the subvolume exists. 
But this error is also getting hit for subvolume which doesn't exist i.e.,for those subvolume which are deleted with "--retain-snapshots" option. Ideally those should be returned with ENOENT.

```
# bin/ceph fs subvolume rm a sub_0 --retain-snapshots
# 
# bin/ceph fs subvolume rm a sub_0 
Error ENOTEMPTY: subvolume 'sub_0' has snapshots
```

ShyamsundarR replied:
If subvolume does not have a current <uuid> directory but has snapshots, it should return ENOTEMPTY, not ENOENT, right? The subvolume is not empty and hence cannot be deleted.

ENOENT would mean subvolume does not exist and further `remove_subvolume_snapshot` understands ENOENT as a success (idempotent remove being a success) and reports it as such to the user.

kotreshhr replied:
I was coming from user's perspective. User might think subvolume and snapshots are different entities. If he has removed subvolume with retaining snapshots, the subvolume no longer exists for him. So I was thinking trying to remove the same subvolume should say "Error ENOENT: subvolume 'xyz' is removed and has only snapshots retained" along the lines of 'subvolume getpath'. 

But I think you made sense there, 'subvolume rm without retain snapshots' should say ENOTEMPTY and 'subvolume rm with retain-snapshot' could be idempotent or say ENOENT.

I am fine with the existing code.

ShyamsundarR commented:
> > > Two generic questions with this PR.
> > > ```
> > > 1. After subvolume deletion, subvolume ls still lists subvolume. Is this expected ?
> > > ```
> > 
> > 
> > Yes, initial thoughts were to not list the subvolume. During implementation found that listing and reflecting the status in info is better, as part of the subvolume is retained.
> 
> I think it's a good idea to mention 'snapshots are retained' along with subvolume names for this case. Users can filter this out if not interested. With current output, it can't differentiated.
> 
> OR
> 
> We can add an option to 'subvolume ls' command to show only subvolumes with retained snaps and not show them by default. This is all from user perspective. What do you think ?

Expanding `subvolume ls` output would be useful, to list volume state along with its name. Makes it easier to determine a subvolume is complete or in snapshot-retained state, than calling info on specific subvolumes.

I did not make any changes here, as we do not open each subvolume to inspect its state at present, and any filtering or addition to ls output would require us to do so. Should we do this now, or wait for a requirement to get the same done?

kotreshhr commented:
> > > > Two generic questions with this PR.
> > > > ```
> > > > 1. After subvolume deletion, subvolume ls still lists subvolume. Is this expected ?
> > > > ```
> > > 
> > > 
> > > Yes, initial thoughts were to not list the subvolume. During implementation found that listing and reflecting the status in info is better, as part of the subvolume is retained.
> > 
> > 
> > I think it's a good idea to mention 'snapshots are retained' along with subvolume names for this case. Users can filter this out if not interested. With current output, it can't differentiated.
> > OR
> > We can add an option to 'subvolume ls' command to show only subvolumes with retained snaps and not show them by default. This is all from user perspective. What do you think ?
> 
> Expanding `subvolume ls` output would be useful, to list volume state along with its name. Makes it easier to determine a subvolume is complete or in snapshot-retained state, than calling info on specific subvolumes.
> 
> I did not make any changes here, as we do not open each subvolume to inspect its state at present, and any filtering or addition to ls output would require us to do so. Should we do this now, or wait for a requirement to get the same done?

I am fine doing this later. I am more inclined not to show subvolumes which has retained snapshots in ls and do it later with an option to show them. 
@vshankar @ajarr @ShyamsundarR thoughts?

ShyamsundarR commented:
Teuthology run results: http://pulpito.front.sepia.ceph.com/shyamsundarr-2020-07-24_20:56:06-fs-shyamsundarr-master-testing-2020-07-24-distro-basic-smithi/

ShyamsundarR commented:
jenkins retest this please

batrick commented on doc/cephfs/fs-volumes.rst:
```diff
@@ -130,16 +130,24 @@ its parent directory and no size limit.
 Remove a subvolume using::
 
-    $ ceph fs subvolume rm <vol_name> <subvol_name> [--group_name <subvol_group_name> --force]
+    $ ceph fs subvolume rm <vol_name> <subvol_name> [--group_name <subvol_group_name> --force --retain-snapshots]
```
nit: the more canonical format would be this:

```suggestion
    $ ceph fs subvolume rm <vol_name> <subvol_name> [--group_name <subvol_group_name>] [--force] [--retain-snapshots]
```

ShyamsundarR replied:
Addressed (also in a few other places in the same rst file)

vshankar commented:
> I skimmed through most of this and the direction looks good to me. Please have @vshankar / @ajarr have another look.

mostly looks good. will give another look tomorrow.

ShyamsundarR commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1650,6 +1920,405 @@ def test_subvolume_rm_with_snapshots(self):
+
+    def test_subvolume_retain_snapshot_clone_from_newer_snapshot(self):
+        """
+        clone a subvolume from recreated subvolumes latest snapshot
```
Done

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -8,18 +8,34 @@
+        - clone snapshot source if subvolume is a clone of a snapshot
+    - It retains backward compatability with legacy subvolumes by creating the meta file for legacy subvolumes under
+    volumes/_legacy/ (see legacy_config_path), thus allowing cloning of older legacy volumes that lack the <uuid>
+    componenet in the path.
```
Done

ShyamsundarR commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -8,18 +8,34 @@
+        - snapshots attached to an ongoing clone operation
+        - clone snapshot source if subvolume is a clone of a snapshot
+    - It retains backward compatability with legacy subvolumes by creating the meta file for legacy subvolumes under
+    volumes/_legacy/ (see legacy_config_path), thus allowing cloning of older legacy volumes that lack the <uuid>
```
Done

kotreshhr commented:
jenkins test make check

batrick commented:
https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20200730.193838

Failures known/unrelated.
